src/async.js

```diff
@@ -1,40 +1,56 @@
-// @ts-check
 import { WalkerBase } from './walker.js';
 
-/** @typedef { import('estree').BaseNode} BaseNode */
-/** @typedef { import('./walker').WalkerContext} WalkerContext */
-
-/** @typedef {(
+/**
+ * @typedef { import('estree').Node} Node
+ * @typedef { import('./walker.js').WalkerContext} WalkerContext
+ * @typedef {(
  *    this: WalkerContext,
- *    node: BaseNode,
- *    parent: BaseNode,
- *    key: string,
+ *    node: Node,
+ *    parent: Node,
+ *    key: string | number | symbol,
  *    index: number
- * ) => Promise<void>} AsyncHandler */
+ * ) => Promise<void>} AsyncHandler
+ */
 
 export class AsyncWalker extends WalkerBase {
 	/**
 	 *
-	 * @param {AsyncHandler} enter
-	 * @param {AsyncHandler} leave
+	 * @param {AsyncHandler} [enter]
+	 * @param {AsyncHandler} [leave]
 	 */
 	constructor(enter, leave) {
 		super();
 
-		/** @type {AsyncHandler} */
+		/** @type {boolean} */
+		this.should_skip = false;
+
+		/** @type {boolean} */
+		this.should_remove = false;
+
+		/** @type {Node | null} */
+		this.replacement = null;
+
+		/** @type {WalkerContext} */
+		this.context = {
+			skip: () => (this.should_skip = true),
+			remove: () => (this.should_remove = true),
+			replace: (node) => (this.replacement = node)
+		};
```

Comment:
These shouldn't need to be here, they are defined by the `super()` calling `WalkerBase`.
But TypeScript + JSDoc seems to be unable to infer these from the `extends WalkerBase` on the class. 🤔 

/cc @wooorm @remcohaszing have you ever run into this before? Do you know of any work arounds?

Reply:
Apparently the culprit are these lines in `visit`:

```js
const _should_skip = this.should_skip;
const _should_remove = this.should_remove;
const _replacement = this.replacement;

// … 

this.should_skip = _should_skip;
this.should_remove = _should_remove;
this.replacement = _replacement;
```

It appears TypeScript is having problems recursively inferring these types from themselves. As a workaround, you can type-cast the values that are later assigned back. Note that this occurs twice!

```js
const _should_skip = /** @type {boolean} */ (this.should_skip);
const _should_remove = /** @type {boolean} */ (this.should_remove);
const _replacement = /** @type {Node | null} */ (this.replacement);
```

---

This is definitely a bug in TypeScript. Here’s minimal reproduction. See what happens if you comment the line `this.name = name`.

```js
class Foo {
  name = ''
}

class Bar extends Foo {
  rename() {
    const name = this.name
    this.name = name
  }
}
```

Reply:
Maybe you need to type fields? Such as like this: https://github.com/vfile/vfile/blob/03efac7cc3a67fe33bb7a0f1de29e17be322549e/lib/index.js#L122-L126

Reply:
No, it’s currently done that way, but it doesn’t make a difference. It’s a bug with subclasses. Also it only happens in JS. If the exact same code in the minimal reproduction is added in a `.ts` file, TypeScript can infer the types perfectly fine.